### PR TITLE
Merge feat 2.0

### LINKED ISSUE #4254: Align the changes to rewards made for "private chain" and for "2.0"
Recently we did a merge of dev to feat-2.0. Because both branches tackled with rewards calculation, there is a TODO added in [era_supervisor.rs](http://era_supervisor.rs/) to remind that we need to verify if we support the compute_rewards chainspec flag correctly in in 2.0. Details [here](https://github.com/casper-network/casper-node/pull/4228#discussion_r1309959599).

TODO:
1. Merge `feat-2.0` into `feat-signature-rewards` (bring data extensibility and private chain changes into rewards feature branch)
2. Check if the private chain changes to rewards work correctly with the rewards changes on the feature branch
3. Optionally (but preferably) merge `feat-signature-rewards` back to `feat-2.0`

### LINKED ISSUE #4316: Version the block header
While merging the feat-2.0 branch to the rewards feature branch, it appeared that versioning the header is mandatory for this to be possible, because there has been some changes done to the header structure, and it's no more the same as the one in V1.

### PR DESCRIPTION
**Pull request by real-felix**
Merge the feature branch with rewarded signatures with 2.0
To make that possible, the block header is now versioned.

### CODE REVIEW

**Comment by rafal-ch** [comment]
bors try

**Review by Fraser999 — APPROVED** [review]

File: Makefile [hunk 1]
```diff
@@ -135,7 +135,7 @@ lint-smart-contracts:
 
 .PHONY: audit-rs
 audit-rs:
-	$(CARGO) audit --ignore RUSTSEC-2022-0093 --ignore RUSTSEC-2023-0044
+	$(CARGO) audit --ignore RUSTSEC-2023-0065
```
Comment (rafal-ch) [inline Makefile]: Instead of this exception, we can now cherry-pick [this](https://github.com/casper-network/casper-node/pull/4312/commits) commit:

**Review by rafal-ch — APPROVED** [review]
LGTM :+1:

**Comment by real-felix** [comment]
There is a problem with the way rewards are calculated. When a block coming from consensus is executed, the rewards must be calculated. When the block comes from the synchronizer, they must not.

I wrote a fix, but it ended up being quite big, so I'd rather merge this PR, and fix this issue in a further one.

**Comment by real-felix** [comment]
I will also fix the `cargo audit` error in that further PR.

**Comment by rafal-ch** [comment]
bors try

File: node/src/components/block_synchronizer.rs [hunk 2]
```diff
@@ -387,8 +387,8 @@ impl BlockSynchronizer {
 
         match &mut self.forward {
             Some(builder) if builder.block_hash() == *block_hash => {
-                if let Some((finalized_block, deploys)) = result {
-                    builder.register_made_finalized_block(finalized_block, deploys);
+                if let Some(executable_block) = result {
+                    builder.register_made_finalized_block(executable_block);
```
Comment (fizyk20) [inline node/src/components/block_synchronizer.rs]: Should the function be renamed to `register_made_executable_block`?

File: node/src/components/block_synchronizer/block_acquisition.rs [hunk 3]
```diff
@@ -119,7 +119,7 @@ pub(super) enum BlockAcquisitionState {
     // with execution are strictly forward sync states. Until a refactor splits
     // the `Complete` states for the historical and forward cases, we need to
     // keep the block around.
-    HaveFinalizedBlock(Box<Block>, Box<FinalizedBlock>, Vec<Deploy>, bool),
+    HaveFinalizedBlock(Box<Block>, Box<ExecutableBlock>, bool),
```
Comment (fizyk20) [inline node/src/components/block_synchronizer/block_acquisition.rs]: `HaveExecutableBlock`?

**Review by fizyk20 — APPROVED** [review]
I reviewed the rewards distribution tests and the `ExecutableBlock` changes and both are looking good, with minor nitpicks :+1: As for the rest, we might have to rely on the tests and fix the issues as they appear :sweat_smile:

File: node/src/components/block_synchronizer/block_builder/tests.rs [hunk 4]
```diff
@@ -239,20 +239,22 @@ fn register_finalized_block() {
         Acceptance::NeededIt
     );
     // Set the builder's state to `HaveStrictFinalitySignatures`.
-    let finalized_block = FinalizedBlock::from(block.clone());
+    //let finalized_block = FinalizedBlock::from(block.clone());
```
Comment (alsrdn) [inline node/src/components/block_synchronizer/block_builder/tests.rs]: nit: stray commented line here

File: node/src/components/block_synchronizer/block_builder/tests.rs
Comment (alsrdn) [inline node/src/components/block_synchronizer/block_builder/tests.rs:1]: nit: rename `register_finalized_block` test with `register_executable_block`

File: node/src/components/contract_runtime.rs [hunk 5]
```diff
@@ -555,17 +614,20 @@ impl ContractRuntime {
                         info!(
                             "ContractRuntime: enqueuing finalized block({}) with {} deploys for execution",
                             finalized_block_height,
-                            deploys.len()
+                            executable_block.    deploys.len()
```
Comment (alsrdn) [inline node/src/components/contract_runtime.rs]: nit: whitespace?

**Review by alsrdn — APPROVED** [review]
Reviewed the `ExecutableBlock` change and it looks good to me with only a new minor naming nits.

File: node/src/components/block_synchronizer.rs (hunk 2 above)
Comment (real-felix) [reply node/src/components/block_synchronizer.rs]: Fixed

File: node/src/components/block_synchronizer/block_acquisition.rs (hunk 3 above)
Comment (real-felix) [reply node/src/components/block_synchronizer/block_acquisition.rs]: Fixed

File: node/src/components/block_synchronizer/block_builder/tests.rs (hunk 4 above)
Comment (real-felix) [reply node/src/components/block_synchronizer/block_builder/tests.rs]: Fixed

File: node/src/components/contract_runtime.rs (hunk 5 above)
Comment (real-felix) [reply node/src/components/contract_runtime.rs]: Fixed

File: node/src/components/block_synchronizer/block_builder/tests.rs
Comment (real-felix) [reply node/src/components/block_synchronizer/block_builder/tests.rs:1]: Fixed

**Comment by real-felix** [comment]
bors merge

**Pull request merged**